modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val validatedList = listOf("1", "22w", "3", "33s").k()
+ *     .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
+ *   //sampleEnd
+ *   println(validatedList)
+ * }
+ * ```
+ *
+ * Notice that in the `Either` case, should any string fail to parse the entire [traverse] is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of [traverse] is closely tied with the [Applicative] behavior of the data type, where computations are runed in isolation.
+ *
+ * Going back to our `IO` example from the beginning with concurrency in mind, we can get an [Applicative] instance for `IO`, by using `parApplicative`. Then when we traverse a `List<A>` with its [Traverse] instance `ListK.traverse()` and a function`(A) -> IO<B>`, we can imagine the traversal as a scatter-gather. Each `A` creates a concurrent computation that will produce a `B` (the scatter), and as the `IO` operations completes they will be gathered back into a `List`.
+ *
+ * Ultimately, we use `parTraverse` that runs each operation in parallel:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.IO
+ * import arrow.fx.extensions.fx
+ * import arrow.fx.extensions.io.concurrent.parTraverse
+ * import arrow.fx.extensions.io.unsafeRun.runBlocking
+ * import arrow.unsafe
+ *
+ * interface Profile
+ * interface User
+ * //sampleStart
+ * data class DummyUser(val name: String) : User
+ * data class DummyProfile(val u: User) : Profile
+ *
+ * fun userInfo(u: User): IO<Profile> =
+ *   IO { DummyProfile(u) } // this can be a call to the DB
+ *
+ * fun List<User>.processLogin(): IO<List<Profile>> =
+ *   parTraverse { userInfo(it) }
+ *
+ * fun program(): IO<Unit> = IO.fx {
+ *   val list = listOf(DummyUser("Micheal"), DummyUser("Juan"), DummyUser("T'Challa"))
+ *     .processLogin().bind()
+ *   println(list)
+ * }
+ * //sampleEnd
+ * fun main() {
+ *   unsafe { runBlocking { program() } }
+ * }
+ * ```
+ *
+ * Evidently, [Traverse] is not limited to [List] or [Nel], it provides an abstraction over 'things that can be traversed over', like a Binary tree, a [SequenceK], or a `Stream`, hence the name [Traverse].
+ *
+ * #### Playing with `Reader`
+ *
+ * Another interesting data type we can use is `Reader`. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> A`.
```

Comment:
```suggestion
 * Another interesting data type we can use is `Reader`. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> Id<A>`.
```
`Kleisli<ForId, D, A>` is just that, you could and probably should add that that is isomorphic to `(D) -> A` tho.